### PR TITLE
Add Knuth-Morris-Pratt (KMP) string matching algorithm [HACKTOBERFEST 2025]

### PR DESCRIPTION
**Pull request by piyushkumar0707**
**Description:**
```markdown
## 🚀 Overview
This PR implements the **Knuth-Morris-Pratt (KMP)** algorithm - one of the most important string matching algorithms that achieves linear time complexity through intelligent preprocessing.

## ✨ Features
- ✅ **O(n + m) time complexity** - Linear performance vs O(n*m) naive approach
- ✅ **Failure function computation** - Detailed preprocessing with visualization
- ✅ **Multiple search variants** - Find first, count all, find all occurrences
- ✅ **Performance benchmarking** - Speed comparison with naive methods
- ✅ **DNA sequence analysis** - Bioinformatics application examples
- ✅ **Educational visualization** - Step-by-step algorithm explanation

## 🎯 Why This Matters
KMP algorithm is fundamental for:
- **Text editors** - Find/replace functionality in IDEs and word processors
- **Bioinformatics** - DNA/RNA sequence analysis and pattern matching
- **Web search engines** - Efficient text indexing and searching
- **Plagiarism detection** - Finding copied text segments
- **Data compression** - Identifying repeated patterns
- **Network security** - Intrusion detection pattern matching

## 📚 Implementation Details
- **Time Complexity:** O(n + m) where n = text length, m = pattern length
- **Space Complexity:** O(m) for the failure function array
- **Preprocessing:** O(m) time to build failure function
- **No backtracking:** Never re-examines text characters

## 🔧 Advanced Functions
- **`compute_failure_function()`** - Core preprocessing with detailed comments
- **`kmp_search()`** - Find all occurrences with position tracking
- **`kmp_search_first()`** - Optimized for finding first occurrence only
- **`kmp_count()`** - Count occurrences without storing positions
- **`visualize_failure_function()`** - Educational visualization tool

## 🧬 Real-World Applications
**DNA Sequence Matching:**
```r
dna_sequence <- "ATCGATCGATCGAATCGATCGATCGAATCGATCG"
pattern <- "ATCG"
matches <- kmp_search(dna_sequence, pattern)
# Finds all genetic marker occurrences

### CODE REVIEW

**Comment by siriak** [comment]
Please check comments

File: string_manipulation/kmp_string_matching.r [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+for (pos in matches2) {
+  match_str <- substr(text2, pos, pos + nchar(pattern2) - 1)
+  cat("Position", pos, ":", match_str, "\n")
+}
+cat("\n")
+
+# Test 3: Edge cases
+cat("3. Edge Cases\n")
+cat("Empty pattern:", length(kmp_search("hello", "")), "matches\n")
+cat("Empty text:", length(kmp_search("", "hello")), "matches\n")
```
[the remark this answers is not in the document]
Comment (siriak) [reply string_manipulation/kmp_string_matching.r:275]: This is still an issue

[the remark this answers is not in the document]
Comment (siriak) [reply string_manipulation/kmp_string_matching.r:275]: Why are you resolving this comment without applying the change or replying to the comment? It's a valid issue that must be fixed

**Review by siriak — CHANGES_REQUESTED** [review]

File: DIRECTORY.md [hunk 2]
```diff
@@ -83,4 +83,9 @@
   * [Shell Sort](https://github.com/TheAlgorithms/R/blob/HEAD/sorting_algorithms/shell_sort.r)
   * [Stooge Sort](https://github.com/TheAlgorithms/R/blob/HEAD/sorting_algorithms/stooge_sort.r)
   * [Topological Sort](https://github.com/TheAlgorithms/R/blob/HEAD/sorting_algorithms/topological_sort.r)
-  * [Wiggle Sort](https://github.com/TheAlgorithms/R/blob/HEAD/sorting_algorithms/wiggle_sort.r)
+    * [Wiggle Sort](https://github.com/TheAlgorithms/R/blob/HEAD/sorting_algorithms/wiggle_sort.r)
+
+## String Manipulation
+  * [KMP String Matching](https://github.com/TheAlgorithms/R/blob/HEAD/string_manipulation/kmp_string_matching.r)
+
+```
```
[the remark this answers is not in the document]
Comment (siriak) [reply DIRECTORY.md]: The issue is indeed there
image

File: string_manipulation/kmp_string_matching.r (hunk 1 above)
[the remark this answers is not in the document]
Comment (piyushkumar0707) [reply string_manipulation/kmp_string_matching.r:275]: @siriak Thank you for the feedback! I've addressed the issues you mentioned:

✅ **Fixed DIRECTORY.md formatting**:
- Corrected the Wiggle Sort indentation (was 4 spaces, now 2 spaces like other entries)
- Cleaned up formatting issues

✅ **Enhanced KMP edge case testing**:
- Added tests for identical strings
- Added tests for pattern at start/end positions  
- Improved output formatting with consistent array display

The formatting and test coverage issues you identified have been resolved. Please review the latest commit.
